src/main/java/com/programmerdan/minecraft/simpleadminhacks/hacks/ShipOutOfLuck.java

```diff
@@ -82,56 +77,44 @@ public void preventBoatPlacements(PlayerInteractEvent event) {
 	}
 
 	@EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = true)
-	public void preventBoatUsage(VehicleMoveEvent event) {
-		Vehicle vehicle = event.getVehicle();
+	public void preventBoatUsage(final VehicleMoveEvent event) {
+		final Vehicle vehicle = event.getVehicle();
 		if (vehicle.getType() != EntityType.BOAT) {
 			return;
 		}
-		List<Entity> passengers = vehicle.getPassengers();
+		final List<Entity> passengers = vehicle.getPassengers();
 		if (passengers.isEmpty()) {
 			return;
 		}
-		Location destination = event.getTo();
-		if (event.getVehicle().isOnGround()) {
+		final Location destination = event.getTo();
+		if (vehicle.isOnGround()) {
 			// If the boat is on top of a block, then we need to look below the boat
 			destination.add(0, -1, 0);
```

Comment:
Won't work for non-full blocks (like slabs)

Reply:
Just tested this and you are correct, but ultimately I'm not sure how to get around this issue. From what I can tell there's no way to know which blocks are in contact with the boat without some presumably some rather expensive bounding box testing. The naive answer to do `-0.5` instead of `-1` but then what if the boat is on trapdoors, carpet, daylight sensors, or <5 layers of snow.. etc?

There's another thing that we might need to worry about, anti-cheat: how robust is it in these regards? We've seen how inventive some hackers can be, even on our server.. so say you have a hacked client that when you ride a boat on ice will transmit packets pretending that you're *above* the ice rather than on it. Will the anti-cheat look below the boat (like what ShipOutOfLuck is doing) and see ice and think the high movement speed is okay?

I dunno, what do you think?

Reply:
Anticheat will eventually catch up to people who send packets pretending they're on ground when they're actually not.

My approach to resolving which block we're actually on would be:
Check if current y coord is less than 1E-6 more than its Math.floor
If yes, use the block below (assume on full block)
If no, use the block the y coord is inside of (assume on non-full block), except if block below is fence or wall

Reply:
Oke, so I've changed tact and gone with a bounding box solution because the incongruous boats wasn't working and would dismount you if you were boating over a block ledge completely unrelated to ice roads. So to alleviate any performance concerns I'm only allowing the check to have 20% of the time, so it'll still happen very frequently because movement events are fired like 20 per second or something.